### ISSUE TITLE
Wrong information for some particles after CabM particle structures rebuild operation

### ISSUE THREAD

**Issue by zhangchonglin** [SCOREC/pumi-pic#119]
While developing on a new particle code, `Comet', I encountered the following issue: 
- If using `CabM` particle structure, when the number of particles per GPU is `>~50 million`, information for some particles are incorrect after particle structure rebuild. 
- When simply switching to `SCS` particle structure, everything works fine.
- This is likely related to the earlier `XGCm` issue when `64 million` particles per GPU was used in that case: https://github.com/SCOREC/xgcm/issues/163.

##### Steps to reproduce the issue:
- In each simulation time steps I generate some new particles and then perform a rebuild operation similar as here: 
https://github.com/SCOREC/pumi-pic/blob/7b55b1b09ec6a30695d45889fd12acf109cdb6be/particle_structs/test/test_rebuild.cpp#L138-L163
- I ensure all the particles information are within a specified (x, y) range (within the simulation domain), for example: `0<=x<=1, 0<=y<=1`.
- But after particle rebuild, many of the particles' position are well outside the simulation domain.

##### Similarly, if no new particles are being added:
- during each time step, update particles' position through push;
- then perform search to find particles' new position/element;
- after that perform a particle structure rebuild;
- If checking particle information after the rebuild, many of the particles have wrong coordinates information: not in the desired element at all. (I assume it will also be true for other particle properties, but coordinates are easier to verify).
- The reason that after rebuild every particle should be in its correct element is that `particle search` will mark particles (which are no longer in the simulation domain) for deletion.
- But apparently, this is not the case, so the most likely reason is some bug in the particle rebuild operation for `CabM` particle structure.

**Comment by zhangchonglin** [comment]
With `CabM` particle structure and only using 1 MPI rank (no migration needed), when the number of particles reaches a value `~>30 million`, the following line is even returning negative number of particles, so definitely some bugs in `CabM` as `num_ptcls` value is already corrupted.
https://github.com/SCOREC/pumi-pic/blob/a352eb9af4ee5375476d59dc54c906ec9a884b74/particle_structs/src/particle_structure.hpp#L65

**Comment by dhyan1272** [comment]
Yes, I see similar issues, the information associated with the particles gets changed and particles are no longer in the element.

**Comment by zhangchonglin** [comment]
Hi Gerrett @diamog: thank you for looking into this. 
- This only happens when the number of particles per GPU is greater than `~40-50` millions, it may be challenging to debug.
- I will see if I can reproduce this with a case having a very coarse mesh but large number of particles. After that I will send that to you.

**Comment by zhangchonglin** [comment]
@diamog: I made a smaller [XGCm test case with 1000 mesh elements](https://github.com/SCOREC/xgc1_data/tree/master/Cyclone_ITG/Cyclone_ITG_deltaf_1000mesh_debug) that can reproduce the error on Summit:
- But this case still uses 4 MPI ranks and 4 GPUs.
- To reproduce the error, it requires more than `50` million particles per GPU.
- With `50` million particles, I did not observe the issue.
- With `55` million particles per GPU, I observe similar issue as in https://github.com/SCOREC/xgcm/issues/163.

```
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
Particle 0 is outside toroidal section [4.712389 < 0.000000 < 6.283185]
```
Looking at this error closer, it corresponds to this code segment, https://github.com/SCOREC/xgcm/blob/1a0ad1e80a4cb099c3fd6b31ab10a1c835aebc4c/src/xgcm_particle.hpp#L329-L344: 
```
    //Check to see if particles are all in correct places
    fp_t major_phi = mesh.getMajorPlaneAngle();
    fp_t minor_phi = mesh.getMinorPlaneAngle();
    auto coords = ptcls->template get<PTCL_COORDS>();
    auto pids = ptcls->template get<PTCL_IDS>();
    auto checkPtcls = PS_LAMBDA(const int& e, const int& p, const bool& m) {
      if (m) {
        auto pid = pids(p);
        if (!is_safe[e])
          printf("Particle %d is in an unsafe element\n", pid);
        if (coords(p,2) < minor_phi || coords(p,2) > major_phi)
          printf("Particle %d is outside toroidal section [%f < %f < %f]\n", pid,
                 minor_phi, coords(p,2), major_phi);
      }
    };
    ps::parallel_for(ptcls, checkPtcls, "check particles");
```
This suggests that `pids` array does not contain correct information any more as all particles' `pid` are 0, in addition to `ptcls->template get<PTCL_COORDS>()` does not contain correct information.

Please let me know if this case is still be too large for your test.

**Comment by diamog** [comment]
Small update, I was able to reproduce a similar failure in pumi-pic's pseudoXGCm. I had to push particle counts to >70 million, but I am seeing particle data getting corrupted. In my case, the mask is getting corrupted so particles "disappear" mid simulation.

I don't have access to Summit, nor xgcm anymore so any reproduction will need to be within pumi-pic. I can work with my current failure and see what I can figure out.

**Comment by zhangchonglin** [comment]
@diamog: Hi Gerrett, thank you for letting us know. I think this is the same issue as I was encountering: particle data is getting corrupted. I think working with pseudoXGCm should be completely fine as you are also seeing this issue when number of particles is very large.

**Comment by Sichao25** [comment]
I was able to reproduce the mask corruption in the SCOREC environment using approximately 70 million particles. Switching from bool to integer for the mask actually resolved the corruption in my test case. I'm unsure whether other bug still persist. I've also reached out to the Cabana team ([here ](https://github.com/ECP-copa/Cabana/issues/842))to get their insights on this problem.
